### PR TITLE
mpfs/corespi: Add driver for CoreSPI

### PR DESCRIPTION
**Pull request by pussuw**
## Summary
Adds a driver for an FPGA fabric / CoreSPI implementation.

Supports multiple instances, assuming they reside in some base address, offsettable by a constant value.

## Impact
MPFS only

### CODE REVIEW

**Comment by pussuw** [comment]
Just for internal review, will upstream and rebase that here instead.

**Review by eenurkka — APPROVED** [review]
Looks good to me, a few nxstyle (add blank line after comment) etc, soon fixed I guess.
BTW, always run cppcheck or some static analysis tool just in case (maybe not included in github tests at the moment)

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 1]
```diff
@@ -0,0 +1,1498 @@
...
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void mpfs_spi_setmode(struct spi_dev_s *dev, enum spi_mode_e mode)
+{
+  /* Mode is selected when the SPI block is being fabricated */
+
+  spierr("Changing SPI mode not supported\n");
```
Comment (jlaitine) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:513]: Err only if setting to something else than the supported mode; I believe some of our SW calls this always (PX4 SPI driver framework) with SPI_MODE_0

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 2]
```diff
@@ -0,0 +1,1498 @@
...
+ * Returned Value:
+ *   none
+ *
+ ****************************************************************************/
+
+static void mpfs_spi_setbits(struct spi_dev_s *dev, int nbits)
+{
+  /* Bitwidth is selected when the SPI block is being fabricated */
+
+  spierr("Changing SPI bitwidth not supported\n");
```
Comment (jlaitine) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:536]: Err only if setting to something else than the supported mode; I believe some of our SW calls this always with 8

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 3]
```diff
@@ -0,0 +1,1498 @@
...
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success. A negated errno value is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+static int mpfs_spi_sem_waitdone(struct mpfs_spi_priv_s *priv)
+{
+  return nxsem_tickwait_uninterruptible(&priv->sem_isr, SEC2TICK(10));
```
Comment (jlaitine) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:637]: Unnecessarily long timeout value?

File: boards/risc-v/mpfs/common/src/mpfs_board_spi.c [hunk 4]
```diff
@@ -95,5 +97,28 @@ int mpfs_board_spi_init(void)
 
 #endif /* CONFIG_SPI_DRIVER */
 #endif /* CONFIG_MPFS_SPI1 */
+
+#ifdef CONFIG_MPFS_CORESPI
+  /* REVISIT: Initialize all of them ? */
```
Comment (jlaitine) [inline boards/risc-v/mpfs/common/src/mpfs_board_spi.c:102]: add a not-implemented #error here if more than 1 corespi is configured, or add initialization of all the configured ones here?

Comment (pussuw) [reply boards/risc-v/mpfs/common/src/mpfs_board_spi.c:102]: We don't use this init file so I deemed it unnecessary to add that here. I can of course do it...

**Review by jlaitine — COMMENTED** [review]
Just some minor comments added, otherwise, LGTM

File: boards/risc-v/mpfs/common/src/mpfs_board_spi.c (hunk 4 above)
Comment (jlaitine) [reply boards/risc-v/mpfs/common/src/mpfs_board_spi.c:102]: you are right, no need

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 3 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:637]: Yes, copy&paste. I'll decrease to 1 sec (which is overkill too). Maybe I'll see if there is something like msec2tick and use some sensible value from that...

File: arch/risc-v/src/mpfs/Kconfig [hunk 5]
```diff
@@ -445,6 +445,36 @@ config MPFS_ETHMAC_1
...
+	default 0x4A000000
+	depends on MPFS_CORESPI
+
+config MPFS_CORESPI_INSTANCES
+	int "Amount of CoreSPI instances"
+	default 1
+	range 1 8
+	depends on MPFS_CORESPI
+
+config MPFS_CORESPI_IRQNUM
```
Comment (jlaitine) [inline arch/risc-v/src/mpfs/Kconfig:470]: What is this/how to use this configuration?

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 1 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:513]: Yes, that is better

File: arch/risc-v/src/mpfs/mpfs_corespi.c (hunk 2 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/mpfs_corespi.c:536]: Ditto

File: arch/risc-v/src/mpfs/Kconfig (hunk 5 above)
Comment (pussuw) [reply arch/risc-v/src/mpfs/Kconfig:470]: It is the interrupt number for CoreSPI. This indicates the first instance interrupt number, others are offset from this by some discrete value (+1 most likely).

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 6]
```diff
@@ -0,0 +1,1544 @@
...
+  status = getreg32(MPFS_SPI_STATUS);
+  if (status & MPFS_SPI_RXOVERFLOW)
+    {
+      mpfs_spi_rxoverflow_recover(priv);
+    }
+
+  DEBUGASSERT(priv->nbits == 8 || priv->nbits == 16);
+
+  priv->fifosize = (MPFS_FIFO_SIZE_BITS / priv->nbits);
+  priv->fifolevel = priv->fifosize / 2;
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:845]: There is no FIFO threshold so I'm not quite sure what to put here.

File: arch/risc-v/src/mpfs/mpfs_corespi.c [hunk 7]
```diff
@@ -0,0 +1,1544 @@
...
+          mpfs_spi_unload_rx_fifo(priv, priv->rxbuf, priv->fifolevel);
+        }
+
+      remaining = priv->rxwords - priv->rx_pos;
+      if (remaining > 0 && rxretry == 0)
+        {
+          /* Last RX character not ready yet, try again (once) */
+
+          rxretry = 1;
+          up_udelay(SPI_TTOA_US(priv->nbits + 2, priv->frequency));
```
Comment (pussuw) [inline arch/risc-v/src/mpfs/mpfs_corespi.c:1288]: This is just hopelessly wrong. The granularity of PCLK_FREQ/DIVIDER is not near enough to use the _requested_ frequency here. The bias is too great. Must use the _actual_frequency.

**Review by jlaitine — APPROVED** [review]
LGTM

**Pull request merged**